### ISSUE TITLE
S1172 FP: Parameter used as extension delegate target

### ISSUE THREAD

**Issue by Luburic** [SonarSource/sonar-dotnet#5338]
### Description

False positive for rule ID S1172 (Unused method parameters should be removed) when the parameter is used as a method group.

### Repro steps

The following code snippet raises the issue on SonarCloud for the smallerOrEqualArray:

        private static double GetDegreeOfOverlap(string[] largerArray, string[] smallerOrEqualArray)
        {
            return (double) largerArray.Count(smallerOrEqualArray.Contains) / largerArray.Length;
        }

### Expected behavior

The rule should not be triggered.

### Actual behavior

The rule is triggered for the smallerOrEqualArray parameter.

### Related information

* SonarCloud for C# .NET 6

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Luburic,

Thank you for reporting this case. I can confirm it as False Positive. 

There must be something very tricky going on under the hood, as we've seen [a similar case](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/MethodParameterUnused.RoslynCfg.NetFx.cs) working under .NET compilation but not working under .NET Framework. This one reproduces under .NET as well.

**Comment by pavel-mikula-sonarsource** [comment]
This has the same root cause as the .NET Framework variant. It's reported to Roslyn here:
https://github.com/dotnet/roslyn/issues/56644

**Comment by DJAxel** [comment]
I have similar issues with the following methods:

```C#
private T GetConfig<T>(string key)
{
    return _config.GetValue<T>($"{_configKey}:{key}");
}

private static string GetFilenameWithoutExtension(string filename)
{
    return filename.Substring(0, filename.IndexOf('.'));
}
```

Not sure if it has the same root cause, but it looks like there is a lit of cases where SonarLint throws in a false positive on S1172. Hopefully my two examples contribute to finding the problem.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @DJAxel, your case is more likely related to the #5491 that was fixed last week and will be part of the next release.

**Comment by mary-georgiou-sonarsource** [comment]
Another post reporting this issue [here](https://community.sonarsource.com/t/csharpsquid-s1172-direct-references-to-methods-of-parameters-are-not-always-recognized-again/85564).

**Comment by m-gallesio** [comment]
To reiterate it here, the problem seems to be only with private methods and not public ones:

```csharp
using System.Collections.Generic;
using System.Linq;

namespace Test;

public class Test
{
    private void BuildSectionPrivate(
        IReadOnlyDictionary<int, string> items, // <- triggers the rule
        IEnumerable<int> values
    )
    {
        var groups = values.Select(items.GetValueOrDefault);
    }

    public void BuildSectionPublic(
        IReadOnlyDictionary<int, string> items, // <- does not trigger the rule
        IEnumerable<int> values
    )
    {
        var groups = values.Select(items.GetValueOrDefault);
    }
}
```

**Comment by mary-georgiou-sonarsource** [comment]
Another similar issue here: https://github.com/SonarSource/sonar-dotnet/issues/7717
```csharp
static void ConfigureServices()
{
    IServiceCollection services = null!;
    IConfiguration configuration = null!;

    void ConfigureOptions<TOptions>(string sectionKey) // <- S1172 sectionKey is not used
        where TOptions : class
    {
        services.Configure<TOptions>(configuration
            .GetSection(sectionKey)
            .Bind);
    }
}
```

Once the delegate `Bind` is added to be passed as a parameter, the `ParameterReferenceOperation` for `sectionKey` disappears from the control flow graph.

**Comment by cristian-ambrosini-sonarsource** [comment]
Another similar issue was reported by [the community](https://community.sonarsource.com/t/method-parameter-reported-as-removable-although-its-being-used/99354):
```cs
private static IEnumerable<DbContext> GetAllDbContextInstances(IServiceProvider serviceProvider) // Remove this parameter 'serviceProvider', whose value is ignored in the method.
{
    var types = GetAllDbContextTypes();
    var instances = types.Select(serviceProvider.GetRequiredService);
    foreach (var instance in instances)
    {
        yield return (DbContext)instance;
    }
}
```

**Comment by m-gallesio** [comment]
Bump, this still applies as of SonarQube 10.4.

**Comment by pavel-mikula-sonarsource** [comment]
Moved to https://sonarsource.atlassian.net/browse/NET-865 (non-public link)

**Comment by chrgrwilde** [comment]
@pavel-mikula-sonarsource : I have seen you moved this issue to a private backlog. How do we know when it will be fixed, especially #5145 we originally filed?

**Comment by pavel-mikula-sonarsource** [comment]
Hi,
It was fixed in 10.4 release in December. The NET-865 is part of the release notes:

https://github.com/SonarSource/sonar-dotnet/releases/tag/10.4.0.108396
